### ISSUE TITLE
null values getting classified and symbolized with numeric attributes 

### ISSUE THREAD

**Issue by makella** [CartoDB/carto-vl#1337]
## Current behavior
Currently, null values are being included in both continuous and classed maps.

In the example dataset, ([poverty by census tract in Brooklyn](https://team.carto.com/u/eschbacher/tables/brooklyn_poverty/)), there are null values in the data that are getting colored as if they have actual numeric values. 

After speaking with @elenatorro yesterday, we did some digging and found that if `null` a value of `0` is returned. In the numeric context, since `0` is actually a value, this isn't the desired behavior.

## Example
In the gif below, you'll see that I highlight some polygons that are getting colored even though they have null values:

![nulls](https://user-images.githubusercontent.com/1566273/57026673-9442a300-6bf7-11e9-9c7f-cbdec50f4b29.gif)

At the end of the gif, you will see that there are also polygon values of 0.

Sample map here: https://bl.ocks.org/makella/bc8837f53a07ea0b6730ccbd342a6308

## Expected behavior
In this context, `null` could be equivalent to: 
- no value so don't show
- or "no data", which is important to be able to highlight on both the map and legend where appropriate:
<img width="695" alt="Screen Shot 2019-05-01 at 10 33 50 AM" src="https://user-images.githubusercontent.com/1566273/57028675-edf99c00-6bfc-11e9-856a-37a3c9860338.png">
 
To satisfy both, default behavior should be similar to category/symbol/etc. ramps where "other values" are sent to an others "category" and colored with our others gray. Similar to those ramps, the other color should have the option to be overwritten, for example, to not symbolize, a user could set the default other color to something like `rgba(0,0,0,0)` or modify our default gray to a color that better suits their map.

cc @andy-esch

**Comment by Jesus89** [comment]
Hi, I have created this nice example in CARTOframes:

![](https://user-images.githubusercontent.com/2227572/58880258-d15af300-86d7-11e9-84d3-f562ca7c82b0.gif)

With this condition, we can visualize null values differently.

So your proposal is to manage this internally in the ramp function, right?

```
// ramp ( input, palette, others, nulls )
ramp(linear($poverty_per_pop), [indigo,orange,white], lightgray, gray)
```

Now I'm wondering why is the `others` color in the ramp if it is not used in the visualization. I would move this param to the `getLegendData()` function and leave the 3rd place for the nulls color.

```
// ramp ( input, palette, nulls )
ramp(linear($poverty_per_pop), [indigo,orange,white], gray)
// and then
color.getLegendData({
    othersLabel: 'Others',
    othersColor: 'lightgray'
});
```

cc @elenatorro @VictorVelarde

**Comment by makella** [comment]
hi @Jesus89 as discussed, my preference would be to keep the model we have for buckets, others, etc.
Basically,
`ramp(linear($poverty_per_pop), [indigo,orange,white])` where we assign our default gray to null values but that can be overwritten:
`ramp(linear($poverty_per_pop), [indigo,orange,white],transparent)` 
Since there won't be an "others" category in the case of numeric data (since colors will interpolate across all values), we can use it in the numeric context for the `null` color.

With that said, what happens when there are nulls in categorical data? As you know, in the case of numeric, the biggest issue was that the 0 value was getting calculated in the map. Since categorical ramps don't accept numbers, what do null values turn into?

Once we know more about that, I think we'll be better informed to make this decision!

**Comment by Jesus89** [comment]
Hi @makella.

This is a good approach, but IMO the critical case is when you have "other" and "nulls" and you want to customize both. If we don't want to break the API, we must keep the "others" position and then add "nulls":

```
// ramp ( input, palette, others, nulls )
ramp(linear($poverty_per_pop), [indigo,orange,white], lightgray, gray) // Both
ramp(linear($poverty_per_pop), [indigo,orange,white], lightgray) // Only others
ramp(linear($poverty_per_pop), [indigo,orange,white], null, gray) // Only nulls
```

It's not super nice but it is consistent with the current API and all the possible cases. It would be nice to get more people involved in this discussion before moving this forward.

**Comment by makella** [comment]
the biggest thing for me is with numeric data there won't be an Others. Only with categorical data. So the "others" for numeric in essence will be null.

Any idea what happens with nulls on categorical data?

**Comment by Jesus89** [comment]
Hi. I have created this example but using categorical data: there are 2 categories + null (Physical, ATM, null). I'm applying red to others value:

**Top 1**
- Physical
- Others (ATM + null)
![](https://user-images.githubusercontent.com/2227572/59213411-1fc73080-8bb5-11e9-9b30-38aee91b813f.png)

**Top 2**
- Physical
- ATM
- Others (null)
![](https://user-images.githubusercontent.com/2227572/59213422-25bd1180-8bb5-11e9-972f-a2d14975d4df.png)

**Top 3**
- Physical
- ATM
- null (empty)
(No others)
![](https://user-images.githubusercontent.com/2227572/59213442-2d7cb600-8bb5-11e9-8b74-cbd1b223b2f0.png)

So in this case, null is treated as another category and it gets its color from the palette.

**Comment by Jesus89** [comment]
In this case, I'm using buckets to select the categories `Physical` and `null`, which maps to an empty string. ATM is in the Others category:

![](https://user-images.githubusercontent.com/2227572/59225863-17312300-8bd2-11e9-80bd-a3d63996d186.png)

**Comment by makella** [comment]
hiii @Jesus89 and @elenatorro 

I did some more reading about representing null values focusing on categorical data. I now better understand that with categories (as much as with numbers) the no data element can be important to represent depending on the topic/data theme. 

With that thinking, @Jesus89 actually, the last example you just posted (where null is actually a category in the data that can be symbolized with buckets) makes most sense to me. This way, if a user wants to specifically color the null values, there is a way to do so, if not, they will get sent to the "others" bucket.
>In this case, I'm using buckets to select the categories Physical and null, which maps to an empty string. ATM is in the Others category.

Here is an example that hit home for me:

- [State of the Polar Bear](http://scimaps.org/mapdetail/state_of_the_polar_b_152). The nulls are identified as "Data Deficient" and in the case of polar bear data was important to show no data as a **political statement to countries like Russia** to release their data:

![State-of-the-polar-bear](https://user-images.githubusercontent.com/1566273/59227237-14aed900-8b92-11e9-99a2-093701d54c62.png)
  - in this case, assuming there are 3 categories + null there are a few different possibilities for symbolization:
  `color: ramp(buckets($trend,["increasing","stable","declining"]), [green,yellow,red])`
  or
  `color: ramp(buckets($trend,["increasing","stable","declining"]), [green,yellow,red],lightgray)`
  or
  `color: ramp(buckets($trend,["increasing","stable","declining",""]), [green,yellow,red,lightgray])`

I think treating the nulls as another category gives the user both the ability to assign specific symbology if/when needed and/or group them with others when that isn't needed.

One question @Jesus89 for the null entry legend above would it be possible to assign a legend label instead of blank?

**Comment by makella** [comment]
Now I'm thinking about other cases..., as you originally suggested, maybe coloring nulls should be an additional parameter:
- In the case where you aren't using buckets, and you color all categories, you wouldn't be able to set a specific color for null categories since every category in the data would be assigned a color:
`ramp($category,bold)`

- To color nulls, we would use your suggestion where `lightgray` is the null (not others since they are all colored) category:
`ramp($category,bold,lightgray)`

BUT, why then, wouldn't you be able to do this for any category? For example, if you wanted to highlight ATMs amongst many categories with a bright yellow color, you wouldn't be able to do that without using `buckets` and listing out every category in the data.

**Comment by elenatorro** [comment]
Nice research!

From @Jesus89 examples, we could return 'Null' by default in `getLegendData()` for the labels. However, it's possible you want to label them differently. In this case, we can add it as an option for the method (as we're already doing for the 'others'). From a CARTO VL perspective that'd be enough, but from CARTOframes we would have to add a parameter to these helpers to allow setting a custom label for the "null category".

In my opinion, since this will require more research / use cases, I think it's a good candidate for the next cycle, because we only have one week left. We could keep the discussion and look for more use cases meanwhile, what do you think?

**Comment by Jesus89** [comment]
Yes. I think we are all aligned now. We know the possible use cases and we need to discuss more to find the best solution.

However, as Elena said, we could prioritize the rest of the tickets for the release (size legends, helpers, docs), and keep this one for the next iteration :+1:

**Comment by makella** [comment]
YEP!!
>In my opinion, since this will require more research / use cases, I think it's a good candidate for the next cycle, because we only have one week left. We could keep the discussion and look for more use cases meanwhile, what do you think?

YEP!!
>However, as Elena said, we could prioritize the rest of the tickets for the release (size legends, helpers, docs), and keep this one for the next iteration 👍

❤️

**Comment by Jesus89** [comment]
This is fixed in the new CF version using WebSDK. It will be available on the next 1.1 release.
